arangod/Aql/SimpleModifier.cpp

```diff
@@ -198,12 +303,35 @@ bool SimpleModifier<ModifierCompletion, Enable>::resultAvailable() const {
 template <typename ModifierCompletion, typename Enable>
 VPackArrayIterator SimpleModifier<ModifierCompletion, Enable>::getResultsIterator() const {
   if (resultAvailable()) {
-    TRI_ASSERT(_results.hasSlice() && _results.slice().isArray());
-    return VPackArrayIterator{_results.slice()};
+    auto const& results = std::get<OperationResult>(_results);
+    TRI_ASSERT(results.hasSlice() && results.slice().isArray());
+    return VPackArrayIterator{results.slice()};
   }
   return VPackArrayIterator(VPackArrayIterator::Empty{});
 }
 
+template <typename ModifierCompletion, typename Enable>
+bool SimpleModifier<ModifierCompletion, Enable>::hasResultOrException() const noexcept {
+  return std::visit(overload{
+                        [](NoResult) { return false; },
+                        [](Waiting) { return false; },
+                        [](OperationResult const&) { return true; },
+                        [](std::exception_ptr const&) { return true; },
+                    },
+                    _results);
+}
+
+template <typename ModifierCompletion, typename Enable>
+bool SimpleModifier<ModifierCompletion, Enable>::hasNoResultOrOperationPending() const noexcept {
+  return std::visit(overload{
+                        [](NoResult) { return true; },
+                        [](Waiting) { return false; },
```

Comment:
Given the method name, shouldn't this case return `true` as well?

Reply:
The name is chosen unwisely. What I meant was, there is neither a result, nor an operation pending. So `NoResult` is the only true case. Perhaps `hasNeitherResultNorOperationPending` would be less ambiguous.

Reply:
Maybe there's also a better name for `NoResult`; e.g. `hasNoResult` could mean both having `NoResult`, or not having `OperationResult`. Perhaps `Nothing` or `Empty`?

Reply:
It is fine then. No need to change.

Reply:
I also it confused by this methods name.
maybe: `hasNothingInFlight` or `hasNothingInProcess` ?
As far as I understood the NoResult indicates: I am ready for more work and not waiting on a response, or someone handling it.
The name `NoResult` and the camel case of `no result` => `NoResult` is causing the ambiguity here.
I actually think the NoResult type is good.

Reply:
I changed the name.